### PR TITLE
Reduce load on electrum

### PR DESCRIPTION
**Pull request by rishflab**
.

### CODE REVIEW

File: swap/src/bitcoin/wallet.rs [hunk 1]
```diff
@@ -202,106 +326,79 @@ impl Wallet {
...
+        let mut inner = self.client.lock().await;
 
-        Ok(BlockHeight::new(height))
+        inner.drain_notifications()?;
+
+        Ok(inner.latest_block)
     }
 
-    pub async fn transaction_block_height(&self, txid: Txid) -> Result<BlockHeight> {
+    pub async fn transaction_block_height(&self, txid: Txid) -> Result<Option<BlockHeight>> {
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:366]: This change could be factored out into a separate commit before this one.

File: swap/src/bitcoin/wallet.rs [hunk 2]
```diff
@@ -66,16 +172,34 @@ impl Wallet {
...
 
+        let electrum = bdk::electrum_client::Client::from_config(electrum_rpc_url.as_str(), config)
+            .map_err(|e| anyhow!("Failed to init electrum rpc client {:?}", e))?;
+
+        let latest_block = electrum.block_headers_subscribe().map_err(|e| {
+            anyhow!(
+                "Electrum client failed to subscribe to header notifications: {:?}",
+                e
+            )
+        })?;
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:183]: Move this into a constructor of `Client`?

File: swap/src/bitcoin/wallet.rs [hunk 3]
```diff
@@ -202,106 +326,79 @@ impl Wallet {
...
-        .context("Transient errors should be retried")?;
 
-        Ok(BlockHeight::new(height))
+        let block_height = reqwest::get(status_url.clone())
+            .await
+            .context("Failed to send request")?
+            .json::<TransactionStatus>()
+            .await
+            .context("Failed to deserialize response as TransactionStatus")?
+            .block_height;
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:381]: Should we actually remove `retry` from this one?

File: swap/src/bitcoin/wallet.rs [hunk 4]
```diff
@@ -202,106 +326,79 @@ impl Wallet {
...
     ) -> Result<()> {
         let conf_target = execution_params.bitcoin_finality_confirmations;
 
         tracing::info!(%txid, "Waiting for {} confirmation{} of Bitcoin transaction", conf_target, if conf_target > 1 { "s" } else { "" });
 
-        // Divide by 4 to not check too often yet still be aware of the new block early
-        // on.
-        let mut interval = interval(execution_params.bitcoin_avg_block_time / 4);
+        self.watch_until_status(txid, script_to_watch, |status| {
+            // TODO: Log the number of confirmations in here to the user.
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:397]: That would be good, need to make sure that we don't repeat it unnecessarily, i.e. only print `1 / 10`, `2 / 10`, etc

Make it follow the format that we have on the Monero side.

File: swap/src/bitcoin/wallet.rs (hunk 3 above)
Comment (rishflab) [reply swap/src/bitcoin/wallet.rs:249]: I will try and replace transaction block height with script status and this can go

File: swap/src/bitcoin.rs [hunk 5]
```diff
@@ -275,3 +272,53 @@ pub struct EmptyWitnessStack;
...
+        let tx_cancel_status = ScriptStatus::Unseen;
+
+        let expired_timelock = current_epoch(
+            CancelTimelock::new(5),
+            PunishTimelock::new(5),
+            tx_lock_status,
+            tx_cancel_status,
+        );
+
+        assert_eq!(expired_timelock, ExpiredTimelocks::Cancel)
```
Comment (thomaseizinger) [inline swap/src/bitcoin.rs:292]: This should be `ExpiredTimelocks::None` I believe.

**Comment by da-kami** [comment]
Summary of test run with @rishflab where we triggered 6 swaps:

No swap finished successfully. ASB gets stuck upon redeeming / monitoring Bitcoin blockchain.

ASB logs:

```
Mar 15 06:15:33 raspberrypi systemd[1]: Started Run the automated swap backend.
Mar 15 06:15:34 raspberrypi asb[1322]: Mar 15 06:15:34.003  INFO swap::trace: Initialized tracing with level: debug
Mar 15 06:15:34 raspberrypi asb[1322]: Mar 15 06:15:34.004  INFO swap::asb::config: Using config file at default path: /var/external/asb/run/asb-config/config.toml
Mar 15 06:15:34 raspberrypi asb[1322]: Mar 15 06:15:34.005 DEBUG swap::database: Opening database at /var/external/asb/run/asb-data/database
Mar 15 06:15:34 raspberrypi asb[1322]: Mar 15 06:15:34.056 DEBUG swap::seed: Reading in seed from /var/external/asb/run/asb-data/seed.pem
Mar 15 06:16:53 raspberrypi asb[1322]: Mar 15 06:16:53.070 DEBUG swap::monero::wallet: Opened Monero wallet asb-wallet
Mar 15 06:16:55 raspberrypi asb[1322]: Mar 15 06:16:55.660 DEBUG swap::kraken::connection: Connected to Kraken websocket API
Mar 15 06:16:55 raspberrypi asb[1322]: Mar 15 06:16:55.842 DEBUG swap::kraken::connection: Subscribed to updates for ticker
Mar 15 06:18:18 raspberrypi asb[1322]: Mar 15 06:18:18.357 DEBUG swap::protocol::alice::event_loop: Connection Established with 12D3KooWScqZscxkgFXRp72WBCg4pHPg4hnZFjAu1vh1pkK5hmVa
Mar 15 06:19:57 raspberrypi asb[1322]: Mar 15 06:19:57.346 DEBUG swap::protocol::alice::event_loop: Connection Established with 12D3KooWFrxKaDB4dc2GXmiQ9RFTTFosHtSizwcxydf3ryvWdMUM
Mar 15 06:20:56 raspberrypi asb[1322]: Mar 15 06:20:56.931 DEBUG swap::protocol::alice::event_loop: Connection Established with 12D3KooWBpsQDwHFAwbKGswsqTXYpPjjQm5wdXZoyaGZrTERShyF
Mar 15 06:21:05 raspberrypi asb[1322]: Mar 15 06:21:05.418  INFO swap::protocol::alice::behaviour: Starting execution setup to sell 0.255830115830 XMR for 0.00099390 BTC peer=12D3KooWScqZscxkgFXRp72WBCg4pHPg4hnZFjAu1vh1pkK5hmVa
Mar 15 06:21:09 raspberrypi asb[1322]: Mar 15 06:21:09.277  INFO swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::protocol::alice::swap: Current state: started
Mar 15 06:21:12 raspberrypi asb[1322]: Mar 15 06:21:12.287 DEBUG swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::bitcoin::wallet: Transaction is unseen txid=41a54395c0a57109870499e779b7fb736db4d54b477ad115ccdd082d1e37b91b
Mar 15 06:21:46 raspberrypi asb[1322]: Mar 15 06:21:46.112  INFO swap::protocol::alice::behaviour: Starting execution setup to sell 0.255830115830 XMR for 0.00099390 BTC peer=12D3KooWFrxKaDB4dc2GXmiQ9RFTTFosHtSizwcxydf3ryvWdMUM
Mar 15 06:21:49 raspberrypi asb[1322]: Mar 15 06:21:49.844  INFO swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::protocol::alice::swap: Current state: started
Mar 15 06:21:50 raspberrypi asb[1322]: Mar 15 06:21:50.034 DEBUG swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::bitcoin::wallet: Transaction is unseen txid=7c16ab19868aa1345b262325e76780f205106379451b2b67b37c63f403d25ab1
Mar 15 06:22:00 raspberrypi asb[1322]: Mar 15 06:22:00.037 DEBUG swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::bitcoin::wallet: Transaction is in mempool txid=7c16ab19868aa1345b262325e76780f205106379451b2b67b37c63f403d25ab1
Mar 15 06:22:00 raspberrypi asb[1322]: Mar 15 06:22:00.037  INFO swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::bitcoin::wallet: Waiting for 1 confirmation of Bitcoin transaction txid=7c16ab19868aa1345b262325e76780f205106379451b2b67b37c63f403d25ab1
Mar 15 06:22:00 raspberrypi asb[1322]: Mar 15 06:22:00.037 DEBUG swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::bitcoin::wallet: Transaction is in mempool txid=7c16ab19868aa1345b262325e76780f205106379451b2b67b37c63f403d25ab1
Mar 15 06:22:00 raspberrypi asb[1322]: Mar 15 06:22:00.136  INFO swap::protocol::alice::behaviour: Starting execution setup to sell 0.255830115830 XMR for 0.00099390 BTC peer=12D3KooWBpsQDwHFAwbKGswsqTXYpPjjQm5wdXZoyaGZrTERShyF
Mar 15 06:22:04 raspberrypi asb[1322]: Mar 15 06:22:04.975  INFO swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::protocol::alice::swap: Current state: started
Mar 15 06:22:05 raspberrypi asb[1322]: Mar 15 06:22:05.168 DEBUG swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::bitcoin::wallet: Transaction is unseen txid=43c7b725a704fa07feafe828c459db77022fa61ce53e38f0f9f2b6bbd8383939
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.669 DEBUG swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::bitcoin::wallet: Transaction is confirmed with 1 blocks txid=43c7b725a704fa07feafe828c459db77022fa61ce53e38f0f9f2b6bbd8383939
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.669  INFO swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::bitcoin::wallet: Waiting for 1 confirmation of Bitcoin transaction txid=43c7b725a704fa07feafe828c459db77022fa61ce53e38f0f9f2b6bbd8383939
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.669 DEBUG swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::bitcoin::wallet: Transaction is confirmed with 1 blocks txid=7c16ab19868aa1345b262325e76780f205106379451b2b67b37c63f403d25ab1
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.669  INFO swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::bitcoin::wallet: Bitcoin tx has 1 out of 1 confirmation txid=7c16ab19868aa1345b262325e76780f205106379451b2b67b37c63f403d25ab1
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.671 DEBUG swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::bitcoin::wallet: Transaction is confirmed with 1 blocks txid=41a54395c0a57109870499e779b7fb736db4d54b477ad115ccdd082d1e37b91b
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.671  INFO swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::bitcoin::wallet: Waiting for 1 confirmation of Bitcoin transaction txid=41a54395c0a57109870499e779b7fb736db4d54b477ad115ccdd082d1e37b91b
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.671 DEBUG swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::bitcoin::wallet: Transaction is confirmed with 1 blocks txid=43c7b725a704fa07feafe828c459db77022fa61ce53e38f0f9f2b6bbd8383939
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.671  INFO swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::bitcoin::wallet: Bitcoin tx has 1 out of 1 confirmation txid=43c7b725a704fa07feafe828c459db77022fa61ce53e38f0f9f2b6bbd8383939
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.672  INFO swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::protocol::alice::swap: Current state: btc is locked
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.672 DEBUG swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::bitcoin::wallet: Transaction is confirmed with 1 blocks txid=41a54395c0a57109870499e779b7fb736db4d54b477ad115ccdd082d1e37b91b
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.672  INFO swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::bitcoin::wallet: Bitcoin tx has 1 out of 1 confirmation txid=41a54395c0a57109870499e779b7fb736db4d54b477ad115ccdd082d1e37b91b
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.675  INFO swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::protocol::alice::swap: Current state: btc is locked
Mar 15 06:22:51 raspberrypi asb[1322]: Mar 15 06:22:51.675  INFO swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::protocol::alice::swap: Current state: btc is locked
Mar 15 06:22:52 raspberrypi asb[1322]: Mar 15 06:22:52.734 DEBUG swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::monero::wallet: sent transfer of 0.255830115830 XMR to 8e4f5c91e6ae5098802c3aaad08f2d0ce707d4adceaa1dda5ea64c4d2e268ca1 in de37bee8db9c8c6c8ff840a248ceb3e2709906cfb45b6247713168f24dbb6f11
Mar 15 06:22:52 raspberrypi asb[1322]: Mar 15 06:22:52.734 DEBUG swap::protocol::alice::behaviour: Sent Transfer Proof
Mar 15 06:22:52 raspberrypi asb[1322]: Mar 15 06:22:52.735  INFO swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::protocol::alice::swap: Current state: xmr is locked
Mar 15 06:22:52 raspberrypi asb[1322]: Mar 15 06:22:52.736 DEBUG swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::bitcoin::wallet: Transaction is confirmed with 1 blocks txid=7c16ab19868aa1345b262325e76780f205106379451b2b67b37c63f403d25ab1
Mar 15 06:22:53 raspberrypi asb[1322]: Mar 15 06:22:53.672 DEBUG swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::monero::wallet: sent transfer of 0.255830115830 XMR to 656f23afa61cd688660df2d6e097654246d4f8285aa3aff82533e2e2b823ffea in f23b5f004b4203f24dba18fb45a5ce385505917854b8235894bd0e019a162f8f
Mar 15 06:22:53 raspberrypi asb[1322]: Mar 15 06:22:53.673 DEBUG swap::protocol::alice::behaviour: Sent Transfer Proof
Mar 15 06:22:53 raspberrypi asb[1322]: Mar 15 06:22:53.674  INFO swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::protocol::alice::swap: Current state: xmr is locked
Mar 15 06:22:53 raspberrypi asb[1322]: Mar 15 06:22:53.675 DEBUG swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::bitcoin::wallet: Transaction is confirmed with 1 blocks txid=43c7b725a704fa07feafe828c459db77022fa61ce53e38f0f9f2b6bbd8383939
Mar 15 06:22:54 raspberrypi asb[1322]: Mar 15 06:22:54.598 DEBUG swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::monero::wallet: sent transfer of 0.255830115830 XMR to e5fd591715d261249e643f62b001150cf6d95dd7555dde083b6becc93eae7bdd in 59665c930a59c719b37399020d8bd94a7dafb38fc797703df2df13360ea6914c
Mar 15 06:22:54 raspberrypi asb[1322]: Mar 15 06:22:54.599 DEBUG swap::protocol::alice::behaviour: Sent Transfer Proof
Mar 15 06:22:54 raspberrypi asb[1322]: Mar 15 06:22:54.599  INFO swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::protocol::alice::swap: Current state: xmr is locked
Mar 15 06:22:54 raspberrypi asb[1322]: Mar 15 06:22:54.600 DEBUG swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::bitcoin::wallet: Transaction is confirmed with 1 blocks txid=41a54395c0a57109870499e779b7fb736db4d54b477ad115ccdd082d1e37b91b
Mar 15 06:24:42 raspberrypi asb[1322]: Mar 15 06:24:42.607 DEBUG swap::protocol::alice::event_loop: Connection Established with 12D3KooWRqVnNWuahKawSSmxQQptGcEApNBFXNodzRv2q6tBvhUW
Mar 15 06:25:39 raspberrypi asb[1322]: Mar 15 06:25:39.585  INFO swap::protocol::alice::behaviour: Starting execution setup to sell 0.255042340261 XMR for 0.00099390 BTC peer=12D3KooWRqVnNWuahKawSSmxQQptGcEApNBFXNodzRv2q6tBvhUW
Mar 15 06:25:42 raspberrypi asb[1322]: Mar 15 06:25:42.866  INFO swap{id=91753eb7-98ba-440a-80c5-a6c7c41f1317}: swap::protocol::alice::swap: Current state: started
Mar 15 06:25:43 raspberrypi asb[1322]: Mar 15 06:25:43.059 DEBUG swap{id=91753eb7-98ba-440a-80c5-a6c7c41f1317}: swap::bitcoin::wallet: Transaction is unseen txid=46e43fd9a8e115d3b6904dd6d02820a9c34cb2825a1a7dcfd0c8fed2e41429ab
Mar 15 06:26:13 raspberrypi asb[1322]: Mar 15 06:26:13.391 DEBUG swap::protocol::alice::event_loop: Connection Established with 12D3KooWB6TRYirguzLdzEyx9piAFN6gvSX4GyekbfM7M5EfhPiU
Mar 15 06:26:39 raspberrypi asb[1322]: Mar 15 06:26:39.381  INFO swap::protocol::alice::behaviour: Starting execution setup to sell 0.255173299101 XMR for 0.00099390 BTC peer=12D3KooWB6TRYirguzLdzEyx9piAFN6gvSX4GyekbfM7M5EfhPiU
Mar 15 06:26:42 raspberrypi asb[1322]: Mar 15 06:26:42.740  INFO swap{id=284af2c5-e25a-4d94-890e-6c5fe3a896bb}: swap::protocol::alice::swap: Current state: started
Mar 15 06:27:04 raspberrypi asb[1322]: Mar 15 06:27:04.041 DEBUG swap::protocol::alice::event_loop: Connection Established with 12D3KooWPVVZPpd3ZFSLnnAX1o1TkMSw4ULeGA9RkVBbez2mDexa
Mar 15 06:27:36 raspberrypi asb[1322]: Mar 15 06:27:36.002  INFO swap::protocol::alice::behaviour: Starting execution setup to sell 0.255173299101 XMR for 0.00099390 BTC peer=12D3KooWPVVZPpd3ZFSLnnAX1o1TkMSw4ULeGA9RkVBbez2mDexa
Mar 15 06:27:39 raspberrypi asb[1322]: Mar 15 06:27:39.209  INFO swap{id=2736ccf4-e16d-45a9-89e7-d08d2a86dc6b}: swap::protocol::alice::swap: Current state: started
Mar 15 06:35:42 raspberrypi asb[1322]: Mar 15 06:35:42.658 DEBUG swap::protocol::alice::encrypted_signature: Received encrypted signature from 12D3KooWScqZscxkgFXRp72WBCg4pHPg4hnZFjAu1vh1pkK5hmVa
Mar 15 06:35:42 raspberrypi asb[1322]: Mar 15 06:35:42.658 DEBUG swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::protocol::alice::steps: Message 3 received, returning it
Mar 15 06:35:42 raspberrypi asb[1322]: Mar 15 06:35:42.659 DEBUG swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::protocol::alice::steps: Message 3 received, returning it
Mar 15 06:35:42 raspberrypi asb[1322]: Mar 15 06:35:42.660 DEBUG swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::protocol::alice::steps: Message 3 received, returning it
Mar 15 06:35:42 raspberrypi asb[1322]: Mar 15 06:35:42.661  INFO swap{id=885156b7-1522-42b3-a37e-3aaba6c0a914}: swap::protocol::alice::swap: Current state: encrypted signature is learned
Mar 15 06:35:42 raspberrypi asb[1322]: Mar 15 06:35:42.663  INFO swap{id=a5af349f-91ad-4a9a-8e1e-51ca50f2aa07}: swap::protocol::alice::swap: Current state: encrypted signature is learned
Mar 15 06:35:42 raspberrypi asb[1322]: Mar 15 06:35:42.664  INFO swap{id=251eceef-3ccf-4cb9-a077-f729a3225d44}: swap::protocol::alice::swap: Current state: encrypted signature is learned
Mar 15 06:35:44 raspberrypi asb[1322]: Mar 15 06:35:44.525 DEBUG swap::protocol::alice::encrypted_signature: Received encrypted signature from 12D3KooWBpsQDwHFAwbKGswsqTXYpPjjQm5wdXZoyaGZrTERShyF
Mar 15 06:35:52 raspberrypi asb[1322]: Mar 15 06:35:52.905 DEBUG swap::protocol::alice::encrypted_signature: Received encrypted signature from 12D3KooWFrxKaDB4dc2GXmiQ9RFTTFosHtSizwcxydf3ryvWdMUM
```

CLI logs (of one of my swaps, same behavior for others): 

```
     Running `target/debug/swap --debug --data-dir /Users/dakami/TEMP/swap/bob/01 buy-xmr --receive-address 53gEuGZUhP9JMEBZoGaFNzhwEgiG7hwQdMCqFxiyiTeFPmkbt1mAoNybEUvYBKHcnrSgxnVWgZsTvRBaHBNXPa8tHiCU51a --seller-addr /ip4/192.168.1.7/tcp/9876 --seller-peer-id 12D3KooWCdMKjesXMJz1SiZ7HgotrxuqhQJbP5sgBm2BwP1cqThi`
2021-03-15 17:24:07  DEBUG Opening database at /Users/dakami/TEMP/swap/bob/01/database
2021-03-15 17:24:07  DEBUG No seed file found, creating at: /Users/dakami/TEMP/swap/bob/01/seed.pem
2021-03-15 17:24:10   INFO Downloading monero-wallet-rpc (65.11 MiB)
2021-03-15 17:24:35  DEBUG Starting monero-wallet-rpc on port 49820
2021-03-15 17:24:42  DEBUG Created Monero wallet swap-tool-blockchain-monitoring-wallet
2021-03-15 17:24:42  DEBUG Requesting quote
2021-03-15 17:24:42   INFO Received quote: 1 XMR ~ 0.00389700 BTC
2021-03-15 17:24:42   INFO Please deposit the BTC you want to swap to tb1qdwqxp9zrcg3ham6kmudy4x36j92kq35w76sn9z (max 0.00500000 BTC)
2021-03-15 17:25:38   INFO Received 0.00100000 BTC
2021-03-15 17:25:38   INFO Spot price for 0.00099390 BTC is 0.255042340261 XMR
2021-03-15 17:25:39  DEBUG Starting execution setup with 12D3KooWCdMKjesXMJz1SiZ7HgotrxuqhQJbP5sgBm2BwP1cqThi
2021-03-15 17:25:43   INFO Published Bitcoin lock transaction txid=46e43fd9a8e115d3b6904dd6d02820a9c34cb2825a1a7dcfd0c8fed2e41429ab
2021-03-15 17:25:46  DEBUG Transaction is unseen txid=46e43fd9a8e115d3b6904dd6d02820a9c34cb2825a1a7dcfd0c8fed2e41429ab
2021-03-15 17:26:44  DEBUG Transaction is confirmed with 1 blocks txid=46e43fd9a8e115d3b6904dd6d02820a9c34cb2825a1a7dcfd0c8fed2e41429ab
2021-03-15 17:26:44   INFO Waiting for Alice to lock Monero
2021-03-15 17:26:44  DEBUG Transaction is confirmed with 1 blocks txid=46e43fd9a8e115d3b6904dd6d02820a9c34cb2825a1a7dcfd0c8fed2e41429ab
```

My analysis so far:

There are several awkward things: the 3 swaps I triggered (the later ones as seen in ASB logs) never got bitcoin confirmations on the ASB (they did on the CLI side...) - generally I feel we are not subscribing correctly here.
For the three swaps that Rishab triggered we do see confirmations and receive the encsig, but then the ASB just seemingly does nothing upon enc-sig-learned, which is when Alice is supposed to redeem.

I think the problem is related to what Thomas said earlier: We have to ensure that we subscribe to looking for tx before publishing it otherwise we might never see it.

File: swap/src/bitcoin/wallet.rs [hunk 6]
```diff
@@ -163,15 +151,15 @@ impl Wallet {
     }
 
     pub async fn get_network(&self) -> bitcoin::Network {
-        self.inner.lock().await.network()
+        self.wallet.lock().await.network()
     }
 
     /// Broadcast the given transaction to the network and emit a log statement
     /// if done so successfully.
     pub async fn broadcast(&self, transaction: Transaction, kind: &str) -> Result<Txid> {
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:159]: I think we should just iterate over all outputs of the transaction here and subscribe to the script!

File: swap/src/bitcoin/wallet.rs [hunk 7]
```diff
@@ -194,6 +190,12 @@ impl Wallet {
             .ok_or_else(|| anyhow!("Could not get raw tx with id: {}", txid))
     }
 
+    pub async fn status_of_script(&self, script: &Script, txid: &Txid) -> Result<ScriptStatus> {
+        let mut client = self.client.lock().await;
+        client.subscribe_to_script(script.clone())?;
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:195]: It doesn't make any sense to subscribe here. The lock is still active so the state cannot possible change between this line and you asking for the status in the next one.

Also, a function named `status_of_script` should return you the status and not have any side-effects!

**Comment by thomaseizinger** [comment]
> I think the problem is related to what Thomas said earlier: We have to ensure that we subscribe to looking for tx before publishing it otherwise we might never see it.

I am not seeing any logs that we publish a transaction at all so something must be going wrong before that already.

**Comment by da-kami** [comment]
> > I think the problem is related to what Thomas said earlier: We have to ensure that we subscribe to looking for tx before publishing it otherwise we might never see it.
> 
> I am not seeing any logs that we publish a transaction at all so something must be going wrong before that already.

Thanks. Bob does publish a Bitcoin transaction though (as seen in his logs above).

File: swap/src/bitcoin/wallet.rs [hunk 8]
```diff
@@ -392,39 +375,37 @@ impl Client {
...
-
-                *history = new_history;
+    fn update_script_histories(&mut self) -> Result<()> {
+        let mut histories = self
+            .electrum
+            .batch_script_get_history(self.script_history.keys())
+            .map_err(|e| anyhow!("Failed to get script histories {:?}", e))?;
+
+        if histories.len() != self.script_history.keys().len() {
+            return Err(anyhow::Error::from(UnexpectedNumberOfScriptHistories {
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:385]: Unless you are planning to name this error type somewhere, e.g. by calling `anyhow::Error::downcast` you can save yourself some typing and just create the error messages directly using `anyhow!` or even `bail!` if you want to also get rid of the explicit `return`.
